src/backend/distributed/commands/multi_copy.c

```diff
@@ -2585,6 +2613,13 @@ CitusCopyDestReceiverShutdown(DestReceiver *destReceiver)
 	List *connectionStateList = ConnectionStateList(connectionStateHash);
 	FinishLocalCopy(copyDest);
 
+	/*
+	 * We are not going to need any reserved connections for this COPY command,
+	 * so proactively deallocate all the reserved connections right now.
+	 */
+	int deallocateCount = copyDest->reserveConnectionPerNode;
+	DeallocateReservedConnections(deallocateCount);
```

Comment:
This still doesn't make a ton of sense to me.

We are somewhat arbitrarily decrementing reservation counts by one, or not, depending on the current reservation count (which may have been affected by nested calls).

Reply:
I think one rule we might want to introduce (at least for now) is that a backend can never reserve more than one connection. We should perhaps separately track whether the reservation was converted into a connection.

Reply:
Ok, I see you point. The inner COPY might get more than one connection and this logic doesn't make sense anymore as the reserved counter gets to  zero for  the  outer  level. 

Let me think a bit if I can find a nicer way.